### ISSUE TITLE
the dataset leaks future information

### ISSUE THREAD

**Issue by siaochuan** [shiyu-coder/Kronos#227]
the dataset leaks future information into every sample via per-window normalization. In finetune_base_model.py (line 127) and finetune_base_model.py (line 132), x_mean/x_std are computed from the full window, which already includes the nominal forecast horizon. Both tokenizer and predictor use this same dataset path, so the leakage contaminates the whole pipeline; see finetune_tokenizer.py (line 20) and finetune_tokenizer.py (line 97). If your goal is unseen-future prediction, this is a stop-ship issue.

**Comment by EthanW1024** [comment]
致命问题，如果这样，模型不可用

**Comment by Lijuncheng75** [comment]
未来信息泄露，是真的吗？

**Comment by CjwlvCyj** [comment]
没仔细看代码，但是这是finetune的模型参数？应该和hugging face上下载到本地的原始预训练模型不一样吧，如果是data leak也只是针对为了做演示的finetune模块？有大神看了原模型可以戳戳

**Comment by njtraveller** [comment]
> 没仔细看代码，但是这是finetune的模型参数？应该和hugging face上下载到本地的原始预训练模型不一样吧，如果是data leak也只是针对为了做演示的finetune模块？有大神看了原模型可以戳戳


一、训练阶段泄露问题的核心证据
1. 数据加载与预处理层的根因（训练 / 微调共用同一套逻辑）
无论是基础训练还是微调，模型均依赖 config.py 中定义的窗口参数（lookback_window=90、predict_window=10）构建样本，且未在训练阶段修正 “全窗口归一化” 的逻辑：

# config.py 核心窗口参数（训练/微调共用）
self.lookback_window = 90  # 历史观测窗口
self.predict_window = 10   # 未来预测窗口
self.window = lookback_window + predict_window + 1  # 样本窗口包含未来数据
训练阶段的数据集构建逻辑与微调完全一致：每个样本的窗口包含「历史段 + 未来段」，且归一化统计量（均值 / 标准差）基于完整窗口（含未来段） 计算，这是泄露的核心根源。

2. Tokenizer 训练阶段的泄露（train_tokenizer.py/finetune_csv/finetune_tokenizer.py）
Tokenizer 是 Kronos 模型的前置训练模块，其训练流程直接依赖污染的数据集：
数据输入层污染：train_tokenizer.py 中 create_dataloaders 加载的数据集与微调共用同一套逻辑，输入的 ori_batch_x 是包含未来预测窗口的归一化数据（已引入未来信息）；
训练目标污染：Tokenizer 的训练目标是 “重构输入数据”（recon_loss = F.mse_loss(z, batch_x)），而输入的 batch_x 是含未来信息的污染数据 ——Tokenizer 学到的是 “含未来信息的特征重构”，而非 “纯历史信息的离散化编码”；
分布式训练不解决泄露：代码中虽使用 DDP 分布式训练，但仅优化训练效率，未改变数据预处理的泄露逻辑。
3. Predictor 训练阶段的泄露（train_predictor.py/finetune_csv/finetune_base_model.py）
Predictor 训练依赖 Tokenizer 输出的 token，而 Tokenizer 已被污染，同时 Predictor 自身训练流程进一步强化泄露：
Token 编码层污染：Predictor 训练时通过 tokenizer.encode(batch_x) 生成 token，而 batch_x 是含未来信息的归一化数据（代码中 batch_x 直接来自数据集，未做去泄露处理）；

# train_predictor.py 核心逻辑（训练阶段）
with torch.no_grad():
    token_seq_0, token_seq_1 = tokenizer.encode(batch_x, half=True)  # batch_x含未来信息
训练目标层污染：Predictor 的训练目标是 “预测下一个 token”（token_out = [token_seq_0[:, 1:], token_seq_1[:, 1:]]），而 token 本身已包含未来信息，导致 Predictor 学到的是 “未来信息驱动的 token 预测”，而非 “历史信息到未来的推理”；
验证逻辑未规避泄露：训练阶段的验证集（val_loader）与训练集使用相同的数据集构建逻辑，验证 loss 是基于泄露数据计算的 “虚假低损失”，无法反映模型真实性能。

**Comment by njtraveller** [comment]
请编写论文和训练模型的大神能回复下，谢谢

**Comment by njtraveller** [comment]
@shiyu-coder 请关注下，谢谢

**Comment by njtraveller** [comment]
经过对代码的仔细分析，该数据集构建过程存在明显的未来信息泄露，主要问题出现在 dataset.py 的实例级归一化（instance normalization）步骤中。具体分析如下：

一、泄露的位置与原因
在 dataset.py 的 __getitem__ 方法中（约第 78–85 行）：

python
# 提取滑动窗口（长度 = lookback_window + predict_window + 1）
win_df = df.iloc[start_idx:end_idx]          # 包含历史 + 预测窗口
x = win_df[self.feature_list].values

# 对整个窗口（包含未来数据）进行均值和标准差计算
x_mean, x_std = np.mean(x, axis=0), np.std(x, axis=0)
x = (x - x_mean) / (x_std + 1e-5)            # 泄露！
窗口定义：self.window = lookback_window + predict_window + 1，即一个样本同时包含 历史输入段 和 未来预测段。

归一化范围：均值和标准差是基于整个窗口（历史+未来）计算的，这意味着：

训练时，模型在预测未来值之前，已经“看到”了未来时间段的全局统计信息（例如未来价格的大幅波动会影响归一化的尺度）。

这会使得模型在推理阶段无法复现相同的归一化条件（因为推理时没有未来的真实值），导致训练与推理的分布不匹配，并严重高估模型性能。
